bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py

```diff
@@ -135,6 +155,492 @@ def legendre_q(lam, mu, x, n_max=20):
+    tau: np.ndarray
+    """tau coordinates of the grid in toroidal coordinates"""
+    sigma: np.ndarray
+    """sigma coordinates of the grid in toroidal coordinates"""
+    tau_c: np.ndarray
+    """tau coordinates of the coils in toroidal coordinates"""
+    sigma_c: np.ndarray
+    """sigma coordinates of the coils in toroidal coordinates"""
+    th_coil_names: list
+    """names of coils to use with TH approximation (always outside the LCFS tau limit)"""
+
+
+def toroidal_harmonic_grid_and_coil_setup(
+    eq: Equilibrium, R_0: float, Z_0: float, radius: float | None = None
+) -> ToroidalHarmonicsParams:
+    """
+    Set up the grid and coils to be used in toroidal harmonic approximation.
+
+    Use the LCFS to find the region over which to approximate psi using TH.
+    Find the coils located outside this region, which can be used in the TH
+    approximation, and find the coils located inside this region.
+
+    Parameters
+    ----------
+    eq:
+        Starting equilibrium to use in our approximation
+    R_0:
+        R coordinate of the focus point in cylindrical coordinates
+    Z_0:
+        Z coordinate of the focus point in cylindrical coordinates
+
+    Returns
+    -------
+    ToroidalHarmonicsParams:
+        Dataclass holding necessary parameters for the TH approximation
+    """
+    # Find region over which to approximate psi using TH by finding LCFS tau limit
+    if radius is None:
+        lcfs = eq.get_LCFS()
+        lcfs_tau, _ = cylindrical_to_toroidal(R_0=R_0, z_0=Z_0, R=lcfs.x, Z=lcfs.z)
+        tau_lcfs_limit = np.min(lcfs_tau)
+    else:
+        x_points = np.array([R_0 - radius, R_0, R_0 + radius, R_0])
+        z_points = np.array([0, radius, 0, -radius])
+        circle_tau, _ = cylindrical_to_toroidal(R_0=R_0, z_0=Z_0, R=x_points, Z=z_points)
+        tau_lcfs_limit = np.min(circle_tau)
+
+    # Using approximate value for d2_min and tau_max to avoid infinities and divide by 0
+    # errors
+    # From the toroidal coordinate transform functions we have that
+    # $\tau = \ln \frac{d_1}{d_2}$, $d_1^2 = (R + R_0)^2 + (z - z_0)^2$, and
+    # $d_2^2 = (R - R_0)^2 + (z - z_0)^2$
+    # We want to approximate the maximum value of tau and the minimum value of d_2.
+    # The maximum value of tau occurs when d_1 is at its maximum value and d_2 is at its
+    # minimum value. d_1 is largest at the focus, where R = R_0 and z = z_0, and so from
+    # the relation above we have that d2_max = 2 * R_0. This gives us that
+    # tau_max = ln(2 * R_0 / d2_min). d_2 is smallest at the focus, and this would give
+    # d_2 = 0. However, this would lead to a divide by 0 error in calculating tau_max
+    # and so we set d2_min to be a small number to avoid this error.
+
+    d2_min = 0.05
+    tau_max = np.log(2 * R_0 / d2_min)
+    n_tau = 200
+    tau = np.linspace(tau_lcfs_limit, tau_max, n_tau)
+    n_sigma = 150
+    sigma = np.linspace(-np.pi, np.pi, n_sigma)
+
```

Comment:
I think this was a previous comment but are these hardcoded numbers made configurable in the followup PR? If not they should be at least be vars with defaults on the function

Reply:
To be addressed in #4006